File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // int maxL = 0;
        // int cL = 0;
     
        // for(int i = 0 ; i<s.size() ; i++){
        //      int hash[255] = {0};

        //      for(int j = i ;j<s.size() ; j++){
        //         if(hash[s[j]] == 1){
        //             break;
        //         }
        //         cL = j-i+1;
        //         maxL = max(maxL , cL);
        //         hash[s[j]] = 1;
        //      }
        // }

        // return maxL;

      int hash[256];
      fill(hash, hash + 256, -1);


       int r = 0 , l = 0;
       int len = 0;
       int maxL = 0;

       while(r < s.size()){
        if(hash[s[r]] != -1){
            if(hash[s[r]] >= l){
                l = hash[s[r]] + 1;
            }
        }
        len = r - l +1;
        maxL = max(len , maxL);

        hash[s[r]] = r;
        r++;

       }

       return maxL;


      

       

       
        
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        st = set()
        l , r = 0 , 0
        max_len = 0
        while r < len(s):

            while s[r] in st and l <= r:
                st.remove(s[l])
                l += 1
            
            st.add(s[r])
            max_len = max(max_len , r-l+1)
            r += 1
        
        return max_len

        